File: app/src/main/java/com/binglivewallpaper/ImageStore.kt
package com.binglivewallpaper

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.io.File

/**
 * Persists the last downloaded Bing image and its metadata so the
 * wallpaper service can render without a network round-trip.
 */
object ImageStore {

    private const val PREFS = "bing_image_store"
    private const val KEY_URL = "url"
    private const val KEY_DATE = "date"
    private const val IMAGE_FILE = "bing_image.jpg"

    fun save(context: Context, bitmap: Bitmap, url: String, date: String) {
        imageFile(context).outputStream().use { out ->
            bitmap.compress(Bitmap.CompressFormat.JPEG, 95, out)
        }
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString(KEY_URL, url)
            .putString(KEY_DATE, date)
            .apply()
    }

    fun load(context: Context): Bitmap? {
        val file = imageFile(context)
        if (!file.exists()) return null
        return BitmapFactory.decodeFile(file.absolutePath)
    }

    fun savedDate(context: Context): String? =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_DATE, null)

    fun savedUrl(context: Context): String? =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_URL, null)

    private fun imageFile(context: Context): File =
        File(context.filesDir, IMAGE_FILE)
}

File: app/src/main/clojure/com/binglivewallpaper/image_store.clj
(ns com.binglivewallpaper.image-store
  (:import [android.content Context]
           [android.graphics Bitmap Bitmap$CompressFormat BitmapFactory BitmapFactory$Options]
           [java.io File FileOutputStream]))

(def ^:private image-filename "bing_image.jpg")

(defn- image-file
  ^File [^Context context]
  (File. (.getFilesDir context) image-filename))

(defn save
  "Persists the bitmap to internal storage."
  [^Context context ^Bitmap bitmap]
  (let [file (image-file context)]
    (with-open [out (FileOutputStream. file)]
      (.compress bitmap Bitmap$CompressFormat/JPEG 95 out))))

(defn- calculate-in-sample-size
  [^BitmapFactory$Options options ^long req-width ^long req-height]
  (let [height (long (.-outHeight options))
        width (long (.-outWidth options))]
    (if (or (> height req-height) (> width req-width))
      (let [half-height (quot height 2)
            half-width (quot width 2)]
        (loop [in-sample-size 1]
          (if (and (>= (quot half-height in-sample-size) req-height)
                   (>= (quot half-width in-sample-size) req-width))
            (recur (* in-sample-size 2))
            (max 1 in-sample-size))))
      1)))

(defn load-image
  "Loads the cached image from disk. If req-width and req-height are provided,
   downsamples to approximately req-width x req-height."
  ([^Context context]
   (let [file (image-file context)]
     (when (.exists file)
       (BitmapFactory/decodeFile (.getAbsolutePath file)))))
  ([^Context context ^long req-width ^long req-height]
   (let [file (image-file context)]
     (when (.exists file)
       (let [path (.getAbsolutePath file)
             bounds-opts (BitmapFactory$Options.)]
         (set! (.-inJustDecodeBounds bounds-opts) true)
         (BitmapFactory/decodeFile path bounds-opts)
         (when (and (pos? (.-outWidth bounds-opts)) (pos? (.-outHeight bounds-opts)))
           (let [sample-size (calculate-in-sample-size bounds-opts req-width req-height)
                 decode-opts (BitmapFactory$Options.)]
             (set! (.-inSampleSize decode-opts) (int sample-size))
             (BitmapFactory/decodeFile path decode-opts))))))))
